capa/features/extractors/ghidra/file.py

```diff
@@ -0,0 +1,221 @@
+                e_lfanew_bytes = e_lfanew_bytes + b
+        except RuntimeError:  # no bytes will be returned, so we can bail out here
+            return
+
+        newoff = struct.unpack("<I", capa.features.extractors.helpers.xor_static(e_lfanew_bytes, i))[0]
+
+        # assume XOR'd "PE" bytes exist within threshold
+        if newoff > MAX_OFFSET_PE_AFTER_MZ:
+            continue
+
+        peoff = off.add(newoff)
+        if seg_max.getOffset() < (peoff.getOffset() + 2):
+            continue
+
+        pe_bytes = b""
+        try:
+            pe_off_bytes = flatapi.getBytes(peoff, 2)
+            for b in pe_off_bytes:
+                b = (b & 0xFF).to_bytes(1, "little")
+                pe_bytes = pe_bytes + b
+        except RuntimeError:
+            return
+
+        if pe_bytes == pex:
+            yield off.getOffset(), i
+
+
+def extract_file_embedded_pe() -> Iterator[Tuple[Feature, Address]]:
+    """extract embedded PE features"""
+
+    for ea, _ in check_segment_for_pe():
+        yield Characteristic("embedded pe"), FileOffsetAddress(ea)
+
+
+def extract_file_export_names() -> Iterator[Tuple[Feature, Address]]:
+    """extract function exports"""
+    st = currentProgram.getSymbolTable()
+    for addr in st.getExternalEntryPointIterator():
+        yield Export(st.getPrimarySymbol(addr).getName()), AbsoluteVirtualAddress(addr.getOffset())
+
+
+def extract_file_import_names() -> Iterator[Tuple[Feature, Address]]:
+    """extract function imports
+
+    1. imports by ordinal:
+     - modulename.#ordinal
+
+    2. imports by name, results in two features to support importname-only
+       matching:
+     - modulename.importname
+     - importname
+    """
+
+    for f in currentProgram.getFunctionManager().getExternalFunctions():
+        for r in f.getSymbol().getReferences():
+            if r.getReferenceType().isData():
+                addr = r.getFromAddress().getOffset()  # gets pointer to fake external addr
+
+        fstr = f.toString().split("::")  # format: MODULE.dll::import / MODULE::Ordinal_*
+        if "Ordinal_" in fstr[1]:
+            fstr[1] = f"#{fstr[1].split('_')[1]}"
+
+        for name in capa.features.extractors.helpers.generate_symbols(fstr[0][:-4], fstr[1]):
+            yield Import(name), AbsoluteVirtualAddress(addr)
+
+
+def extract_file_section_names() -> Iterator[Tuple[Feature, Address]]:
+    """extract section names"""
+
+    for block in currentProgram.getMemory().getBlocks():
+        yield Section(block.getName()), AbsoluteVirtualAddress(block.getStart().getOffset())
+
+
+def extract_file_strings() -> Iterator[Tuple[Feature, Address]]:
+    """extract ASCII and UTF-16 LE strings"""
+
+    for block in currentProgram.getMemory().getBlocks():
+        p_bytes = b""
+        addr = block.getStart()
+        while block.isInitialized() and addr.getOffset() <= block.getEnd().getOffset():
+            p_bytes = p_bytes + ((block.getByte(addr) & 0xFF).to_bytes(1, "little"))
```

Comment:
See earlier comment about creating a helper function that combines “getBytes” and the signed to unsigned conversion.

Also, can we grab all of the bytes from a block using a single call to “getBytes” e.g. “getBytes(block.getStart(), block.getEnd() - block.getStart())”?